### PR TITLE
improvements/typo fixes + (blank) profile pic

### PR DESCRIPTION
**Pull request by ghost**
urls are now unshortened;
search_terms are unquote_plus'ed whitin the search box
add timestamp when scraping handle
better logic regarding first_run

### CODE REVIEW

File: twatbot.py [hunk 1]
```diff
@@ -493,7 +493,7 @@ def fetch_more_tweets_callback(user, twats):
 
 def scrape(user, first_run = False):
 
-	if first_run and not args.count == -2:
+	if first_run and not os.path.isfile('users/%s/twats.json' % (user.lower())):
```
Comment (rofl0r) [inline twatbot.py]: don't we have a function like get_twats_filename ?

Comment (rofl0r) [reply twatbot.py]: i think this should be if first_run and (args.count != -2 or not fileexists(...))

File: twatbot.py [hunk 2]
```diff
@@ -500,11 +500,11 @@ def scrape(user, first_run = False):
 		checkfn = fetch_more_tweets_callback
 		count = -1
 
+	elapsed_time = time.time()
 	insert_pos = 0
-	sys.stdout.write('\rscraping %s... ' % user)
+	sys.stdout.write('\r[%s] scraping %s... ' % (time.strftime("%Y-%m-%d %H:%M:%S", time.gmtime(elapsed_time)), user))
```
Comment (rofl0r) [inline twatbot.py]: please... a function get_timestamp(date=None) that returns a string
if date is None date = time.time()

File: twatbot.py [hunk 3]
```diff
@@ -207,18 +207,32 @@ def user_at_link(user):
...
-
-	if os.path.isfile('users/%s/profile.jpg' % twat['owner'].lower()):
-		tw += '<div class="profile_picture"><img width="100%%" height="100%%" src="users/%s/profile.jpg"></div>' % twat['owner'].lower()
+	tweet_pic = None
+	retweet_pic = None
 
 	if not 'rid' in twat:
 		retweet_str = ""
+		if has_profile_pic(twat['owner']): tweet_pic = 'users/%s/profile.jpg' % twat['owner'].lower()
+		else: tweet_pic = None
```
Comment (rofl0r) [inline twatbot.py]: tweet_pic is already None

File: twatbot.py [hunk 4]
```diff
@@ -207,18 +207,32 @@ def user_at_link(user):
...
 	tw = '<div class="twat-container">'
-
-	if os.path.isfile('users/%s/profile.jpg' % twat['owner'].lower()):
-		tw += '<div class="profile_picture"><img width="100%%" height="100%%" src="users/%s/profile.jpg"></div>' % twat['owner'].lower()
+	tweet_pic = None
+	retweet_pic = None
 
 	if not 'rid' in twat:
 		retweet_str = ""
+		if has_profile_pic(twat['owner']): tweet_pic = 'users/%s/profile.jpg' % twat['owner'].lower()
```
Comment (rofl0r) [inline twatbot.py]: can we haz a function get_profile_pic_path()

File: twatbot.py [hunk 5]
```diff
@@ -729,6 +748,9 @@ def load_watchlist():
 
 	load_watchlist()
 
+	if args.unshorten:
+		with open('shorteners.txt', 'r') as f: shorteners = [ i.strip() for i in f.read().split('\n') if len(i.strip()) ]
```
Comment (rofl0r) [inline twatbot.py]: is that being done before the chdir() used by --dir ?

File: twat.py [hunk 6]
```diff
@@ -28,6 +28,19 @@ def _hash(str):
...
+def _get_real_location(url, proxies=None):
+	url_components = _split_url(url)
+
+	http = RsHttp(url_components['host'], ssl=url_components['ssl'], port=url_components['port'], keep_alive=True, follow_redirects=True, auto_set_cookies=True, proxies=proxies, user_agent="curl/7.60.0")
+
+	if not http.connect(): return url
+	hdr = http.head(url_components['uri'])
+
+	for line in hdr.split('\n'):
+		if line.lower().startswith('location:'): return line.split(' ')[1].strip()
```
Comment (rofl0r) [inline twat.py]: shouldnt that be 'location: ' and split(': ', 1)

File: twat.py [hunk 7]
```diff
@@ -102,6 +115,19 @@ def _mirror_file(url_components, user, tid, args=None, content_type=None, force=
 	if os.path.lexists(outname): os.unlink(outname)
 	os.symlink('../../data/%s.%s' % (filehash, ext), outname)
 
+def _unshort_urls(twat, proxies=None, shorteners=[]):
```
Comment (rofl0r) [inline twat.py]: why is that func name _ prefixed, when it is supposed to be a public API ?

File: twatbot.py [hunk 8]
```diff
@@ -207,18 +208,33 @@ def user_at_link(user):
 		return '<a href="?user=%s">@</a>' % user
 	return '<a href="https://twitter.com/%s">@</a>' % user
 
+def has_profile_pic(user):
+	return os.path.isfile('users/%s/profile.jpg' % user.lower())
```
Comment (rofl0r) [inline twatbot.py]: shoulnt that func use get_profile_pic_path() too ?

File: twat.py [hunk 9]
```diff
@@ -102,6 +115,19 @@ def _mirror_file(url_components, user, tid, args=None, content_type=None, force=
 	if os.path.lexists(outname): os.unlink(outname)
 	os.symlink('../../data/%s.%s' % (filehash, ext), outname)
 
+def unshort_urls(twat, proxies=None, shorteners=[]):
```
Comment (rofl0r) [inline twat.py]: i think "unshorten" is the proper term to use here

File: twat.py [hunk 10]
```diff
@@ -276,8 +276,17 @@ def get_style_tag(tag, styles):
 		if tg.strip() == tag: return s.strip()
 	return None
 
-def fetch_profile_picture(html, user, proxies):
-	soup = soupify(html)
+def fetch_profile_picture(user, proxies, res=None, twhttp=None):
+	user = user.lower()
+	if not os.path.isdir('users/%s' % user):
```
Comment (rofl0r) [inline twat.py]: don't we have a getuserdir() func or so

Comment (rofl0r) [reply twat.py]: actually i think we should move the make-dir-if-not-exist logic out of that func

File: twat.py [hunk 11]
```diff
@@ -250,8 +272,15 @@ def get_style_tag(tag, styles):
 		if tg.strip() == tag: return s.strip()
 	return None
 
-def fetch_profile_picture(html, user, proxies):
-	soup = soupify(html)
+def fetch_profile_picture(user, proxies, res=None, twhttp=None):
+	if not os.path.isdir('users/%s' % user.lower()): os.makedirs('users/%s' % user.lower())
```
Comment (rofl0r) [inline twat.py]: paths.get_user

File: twat.py [hunk 12]
```diff
@@ -315,10 +344,14 @@ def extract_twat(soup, twats, timestamp):
 
 			tweet_id = div.attrs["data-tweet-id"]
 			tweet_user = div.attrs["data-screen-name"]
+			# fetch profile picture
+			fetch_profile_picture(tweet_user, proxies, twhttp=twhttp)
```
Comment (rofl0r) [inline twat.py]: why is this in extract_twat ? extract_twat is the func thats called to extract _a single tweet_ from the html
i.e. its called 20x per page

File: twatbot.py [hunk 13]
```diff
@@ -714,6 +735,13 @@ def load_watchlist():
 	if args.mirror_size > 0:
 		args.mirror_size = args.mirror_size * 1024*1024
 
+	if args.unshorten:
+		shorteners = {}
```
Comment (rofl0r) [inline twatbot.py]: why is this conditional on args.unshorten ?

**Comment by ghost** [comment]
as this PR was broken, r0fl0r helped fixing it, and another PR has been created

**Pull request closed without merging**